### ISSUE TITLE
[ENH] reduce boilerplate in data set loaders, introduce tests

### ISSUE THREAD

**Issue by fkiraly** [alan-turing-institute/sktime#2697]
Currently, there is a lot of repetitive boilerplate in the `datasets` modules, and no tests for individual loaders such as `load_PBS_dataset`, `load_japanese_vowels`, etc

What should be done here:

* reduce boilerplate, refactor common code into single location
* ensure function signatures are the same, deprecate if needed
* add an argument `return_mtype` such as in the synthetic data generators, that allows users to specify a data format, and uses `datatypes.convert` internally (this is also boilerplate and can go into a joint location)
* ensure the datasets are listed in the documentation (and potentially retrievable using a runtime registry utility)
* there should be tests for all the specific loaders, I think currently there are none

**Comment by GigaMoksh** [comment]
Hey, I'll like to work on this issue. I won't be able to do all but I can definitely try!

**Comment by GigaMoksh** [comment]
I went through the modules in the `datasets` directory. You want to reduce the boilerplate from `_data_io.py`, `_single_problem_loaders.py`, `setup.py` and `tsc_dataset_names.py` files right?

**Comment by fkiraly** [comment]
@GigaMoksh, yes. Good spot on `setup.py`, I think that could also be compressed to a registry-like lookup.
I was mostly concerned about repetition and multiplication of boilerplate in `_single_problem_loader` and `_data_io`.

There are some straightforward refactor steps that clean up boilerplate, and some "nice to have" along the above lines. Every little bit helps, so much appreciated even if you don't cover everything!

**Comment by GigaMoksh** [comment]
Okay so do you want to create a new file for all those boilerplate code?

**Comment by GigaMoksh** [comment]
Also, can you please point out where the code is being repeated? Because I can't see where the code is being repeated. Thanks

**Comment by fkiraly** [comment]
There´s repetition in the loaders, starting with `load_` etc, kindly have a look.

The best thing would be if you could go through the code and propose a target structure for the refactor before implementing so we can discuss/refine.

**Comment by achieveordie** [comment]
Hello @fkiraly, I want to contribute to sktime and since this was marked as a "good first issue" and I'm a newcomer, I made a few changes in the structure. 

- In the `_data_io.py` file, I've refactored the private functions `_load_provided_dataset()` and `_load_dataset()` in a manner that `_load_dataset()` takes care extraction of data and then calls the former to load the provided data. This refactor reduces the boilerplate code of loading data from the given split.

- My concern initially was the inconsistent use of `load_from_tsfile_to_dataframe()` and `load_from_tsfile()` in both functions, so in the refactored version, I've only made use of `load_from_tsfile()` as the sole entrypoint since that seems to call `load_from_tsfile_to_dataframe()` anyways and gives us more control over the datatype. 

- These changes also require the addition of a few default parameters into both functions, like having a parameter `return_type=None` in `_load_dataset()` and having parameters `local_module` and `local_dirname` in `_load_provided_dataset()`, with the default values as the global variable `MODULE` and `DIRNAME` respectively.

I haven't read the specific file loader functions line by line, but to me, it seems like they shouldn't be refactored as they parse different formats differently, so I wanted to know your opinion on it.

**Comment by fkiraly** [comment]
>Hello @fkiraly, I want to contribute to sktime and since this was marked as a "good first issue" and I'm a newcomer, I made a few changes in the structure. 

@achieveordie, this all looks like very sensible changes! I would probably also identify those as primary refactor points.

Could you kindly open a pull request? Let us know if you need assistance how to do that (fork, then PR)
https://www.sktime.org/en/stable/developer_guide/git_workflow.html

**Comment by achieveordie** [comment]
I've dropped in a pull request as you mentioned, thanks for the link. 

I think I've found a bug but since I'm not sure I'm writing it here as a comment rather than opening an issue. In the case where `return_X_y` is `False`, the code appends the response variable into predictors using the column name 'class_var'. But this is only possible if the predictor variable X is an instance of a DataFrame and should fail in the case where X is a 2D or 3D numpy array. Am I missing something?

**Comment by fkiraly** [comment]
> But this is only possible if the predictor variable X is an instance of a DataFrame and should fail in the case where X is a 2D or 3D numpy array. Am I missing something?

Hm, sounds like a bug!

Could you kindly open a bug report (new issue) and post a self-contained snippet of code that provokes the bug?
